Code quality inspection for Clojure using SonarQube (#314)
solita/solita.github.com: _posts/2019-03-08-code-quality-inspection-for-clojure-with-sonarqube.md

---
layout: post
title: Code quality inspection for Clojure using SonarQube  
author: hjhamala
excerpt: Clojure has good libraries for code quality analyzing and vulnerabilities detection. I introduce the support for Kibit, ancient-clj, lein-nvd and Cloverage in existing Clojure SonarQube plugin. 
tags:
- SonarQube
- Clojure
- ClojureScript
- Code quality
- Vulnerabilities detection
- Continuous inspection
---

> Building good and secure code has never been easy. Programmers make mistakes during development. Programs use external libraries which may have  bugs and vulnerabilities. New vulnerabilities are found which forces updates to existing programs.

Reviewer: sanoisin: ...may contain bugs...

Ehkä myös jotain:
New vulnerabilities are revealed constantly which enforces patching existing programs.

Reviewer 2: I think "forces" is more correct than "enforces" here.

> Building good and secure code has never been easy. Programmers make mistakes during the development. Programs use external libraries which may have  bugs and vulnerabilities. New vulnerabilities are found which forces to update existing programs.

Reviewer 2: "during development"
"forces updates to" or something like that

> Software development have many ways to battle against bugs. For example testing, code reviews and pair programming can help us to make better and safer programs. Fortunately we may also use code analyzers which automate the testing and vulnerability finding.

Reviewer 2: "development has" or "developers have"

> [**Clojure**](https://clojure.org/) is not immune for bugs or vulnerabilities. Clojure as a dynamic language leads to situations where some problems may occur in runtime which in static languages are found out in compile time. The Clojure community have produced many useful tools such as linters and security scanners.

Reviewer 2: "immune to"
"Clojure being a dynamic language" or some restructuring
"occur at runtime"
"found out at compile time" or "found during compiling" or something similar
"community has produced"

## Clojure code quality tools

### Eastwood

> [Eastwood](https://github.com/jonase/eastwood) is a lint tool for Clojure (no CLJS support) which detects many kinds of potentially problematic code. 

Reviewer: ```suggestion
[Eastwood](https://github.com/jonase/eastwood) is a linting tool for Clojure which detects many kinds of potentially problematic code. Unfortunately, it lacks ClojureScript support.
```

> For example next Clojure code contains a potential bug:

Reviewer 2: "For example the following Clojure code"

> For example the following Clojure code contains a potential bug:

Reviewer: For an example

Reviewer 2: https://ell.stackexchange.com/questions/19923/for-example-vs-for-an-example

```clojure
(if (and x)
  (foo))
```

> When running ```lein eastwood``` it produces next result:

Reviewer 2: "produces the following result"

> When running ```lein eastwood``` it produces next result:

Reviewer 2: Not sure how the triple tick quoted (```) inline code is displayed here. I'd guess using a single tick is more correct for inline code in any case.

(This occurs a few times below as well.)

Reviewer: produces following result:

```
== Eastwood 0.3.3 Clojure 1.8.0 JVM 11.0.1
Directories scanned for source files:
 src test
== Linting clojure-sonar-example.core ==
Entering directory `/Users/heikkiha/src/sonarqube/clojure-sonar-example'
src/clojure_sonar_example/core.clj:3:5: suspicious-expression: and called with 1 args.  (and x) always returns x.  Perhaps there are misplaced parentheses?
```
 

> Eastwood may cause false positives. Fortunately it is possible to suppress linter rules which may cause any false positives.

Reviewer 2: "may find false positives"

Reviewer: Laittaisin: Fortunately, it is...

 
### Kibit

> [Kibit](https://github.com/jonase/eastwood) is a static code analyzer which detects code that could be rewritten with a more idiomatic function or macro. It supports also ClojureScript. 

Reviewer: Väärä linkki?

can be rewritten? Ehkä myös a macro (en ole ihan saletti, tuleeko artikkeli)

Example: 

```clojure
(> x 0) 
```

```lein kibit``` produces next result:

```
At src/clojure_sonar_example/core.clj:13:
   Consider using:
     (pos? x)
   instead of:
     (> x 0)
```

> I think Kibit is most useful for beginning Clojure programmers but it contains some good advices for more experienced developers also. 

Reviewer 2: "good advice"

### Cloverage

> [Cloverage](https://github.com/cloverage/cloverage) is a code coverage tool for Clojure which runs the tests ot a program 

Reviewer 2: "tests o*f* a program"

and calculates line and form coverage for namespaces. 

Sample output looks like this:

|                    Namespace | % Forms | % Lines |
|------------------------------|---------|---------|
|foo.bar                       |   70.00 |   84.62 |
|foo.baz                       |   94.61 |   96.89 |
|foo.config                    |   60.71 |   80.65 |
|foo.router                    |   31.75 |   69.62 |
|foo.login                     |   68.67 |   82.76 |
|foo.util                      |    2.89 |   20.00 |
|------------------------------|---------|---------|
|                    ALL FILES |   67.36 |   82.93 |

### Lein-nvd

> [Lein-nvd](https://github.com/rm-hull/lein-nvd) is a dependency-checker plugin whichs checks JARS in the programs classpath for known vulnerabilites against 

Reviewer 2: "in the program's classpath"

the [National Vulnerability Database](https://nvd.nist.gov/). 

> Output of Lein-nvd gives status of the JARs with found vulnerabilities:

Reviewer: with known vulnerabilities

| dependency             | status|
|------------------------|-------|
| bcpkix-jdk15on-1.58.jar | OK    | 
| bcprov-jdk15on-1.58.jar| CVE-2017-13098, CVE-2018-1000180, CVE-2018-1000613|

> It should be noted that Lein-nvd does not exactly tell which dependency or its transitive dependency pulls a vulnerable JAR file. Most time JAR name matches a dependency. The full dependency tree can be inspected by running ```lein deps :tree``` command.

Reviewer 2: "does not tell exactly which"
"direct or transitive dependency"
"running *the* `lein deps :tree` command"

### Ancient

[Ancient-clj](https://github.com/xsc/lein-ancient/tree/master/ancient-clj) is a plugin to check your project for outdated dependencies and plugins and  suggest updates. 

Sample output looks like this:

```
[ring "1.7.1"] is available but we use "1.6.3"
```

> Ancient can be started with command ```lein ancient upgrade :interactive``` which gives option to one by one decide if the upgrade is made or not. The upgrade also runs tests and persists the changes only if the tests run correctly. 

Reviewer 2: "gives the option"
"to decide one by one if"

Ancient cannot analyze if the new version is backwards compatible or not. Many libraries uses semantic versioning major.minor.patch (1.2.3) where minor (1.**2**.3) and patch (1.1.**2**) version changes should be backwards compatible. Whether this is really the case can differ.

## Consolidating code quality and vulnerability findings using SonarQube

> [SonarQube](https://www.sonarqube.org/) is a code inspection tool which supports on the box many currently used languages like Java, Python, C++ and so on. SonarQube server have a browser GUI for browsing the findings.

Reviewer 2: "supports out of the box"
Though perhaps "supports many ... out of the box" would be even better.
"server has a browser GUI"

![SonarQube GUI](/img/code-quality-inspection-for-clojure-with-sonarqube/sonar-results.png)

> SonarQube unfortunately doesnt have out of a box support for Clojure. Fortunately SonarQube can be extended to support more languages with plugins. 

Reviewer 2: "doesn't"
"out of the box support"

> There have been many different Clojure plugins for SonarQube but most of them have not been maintained and doesnt work with newer versions of SonarQube. When I started to look for SonarQube plugin only working one was [plugin](https://github.com/fsantiag/sonar-clojure) made by Felipe Santiago. The plugin only supported Eastwood based linting.

Reviewer 2: "and don't work"
"the only working one was the plugin"

> Because plugin was open sourced and the maintainer was active I made necessary changes to the plugin which added support for Kibit, Cloverage, Lein-nvd and Ancient which are now merged. The plugin have also a minimal SonarQube docker image file for testing the plugin locally. 

Reviewer 2: "Because the plugin"
"made the necessary changes"
"The plugin also has a minimal"

### Running SonarQube locally
* Install Docker if not installed earlier
* Clone the source ```git clone https://github.com/fsantiag/sonar-clojure```
* Run ```start-sonarqube.sh``` which builds the plugin with Maven, creates a new docker SonarQube image with plugin copied on to it and finally starts the SonarQube.
* Open http://localhost:9000 to check that SonarQube works correctly. Use admin/admin as the administrator password.
* Download SonarQube client from [https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner](https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner) and install it to path.

### Changes to Clojure application

> Use already existing application or clone a sample application with existing vulnerabilities and code smells ```git clone https://github.com/hjhamala/sonar-clojure-example```.

Reviewer 2: "Use *an* already existing"

Create a sonar-project.properties file in the root folder of your app:

```properties
sonar.projectKey=your-project-key
sonar.projectName=YourProjectName
sonar.projectVersion=1.0
sonar.sources=src,project.clj
sonar.clojure.lein-nvd.json-output-location=target/nvd/dependency-check-report.json
sonar.clojure.cloverage.json-output-location=target/coverage/codecov.json
```

> Add next dependencies to project.clj in project.clj or user profile file:

Reviewer 2: "Add the following dependencies"
"project.clj" is duplicated

> Add next dependencies to project.clj or user profile file:

Reviewer: following dependencies

```clojure
:plugins [[lein-ancient "0.6.15"]
          [jonase/eastwood "0.3.3"]
          [lein-cloverage "1.0.13"]
          [lein-kibit "0.1.6"]
          [lein-nvd "0.6.0"]]
```

Run ```sonar-scanner``` in the root folder of the project.

Open a web browser in http://localhost:9000/dashboard?id=your-project-key

You should see something like this:

![Example Dashboard](/img/code-quality-inspection-for-clojure-with-sonarqube/example-dashboard.png)

Opening project.clj shows vulnerability findings and outdated dependency warning:

![Project vulnerabilities](/img/code-quality-inspection-for-clojure-with-sonarqube/project-vulnerabilities.png)

> Core.clj have findings from Kibit, Eastwood and code coverage from Cloverage:

Reviewer 2: "Core.clj has findings"

![Bad Smell Code](/img/code-quality-inspection-for-clojure-with-sonarqube/smell.png)

## Limitations of the plugin

The SonarQube plugin currently marks all the found vulnerabilites to the first line of project.clj because the lein-nvd only returns JAR name not the dependency which pulls it directly or transitively.

> Only line coverage is supported by SonarQube and calculation seems to be be somehow different compared to Cloverage itself by few percents.

Reviewer 2: "and the calculation"
doubled "be"
"to Cloverage by a few percentage points"

Clojure language parser for Sonarcube has not been implemented. This means that for example syntax highlighting does not currently work. I have a work in progress for adding the [syntax highlighting](https://github.com/hjhamala/sonar-clojure/tree/syntax-highlight-and-tokenization) and hopefully this can be added to the plugin very soon.

## Future improvements

> Current version of the plugin is a native Java plugin which invokes Leiningen to analyze source code once per sensor. Running Leiningen takes few seconds per a sensor. Instead of invoking Leiningen we could make a native Clojure plugin which includes analyzing libraries. This would make running analyzes very much faster because multiple Leiningen invocations are not needed any more. Also the need to install Leiningen plugins with correct versions is no more needed. 

Reviewer 2: "includes the analyzing libraries"
"running the analysis"

> Next improvements should also be considered as well:

Reviewer: Following improvements

* Cloverage form coverage could be added as custom metric to files. 
* Lein-nvd vulnerabilities could be mapped to correct dependency and namespaces requires. This can be made by generating full dependency tree and trying to match the JAR name with dependencies.

> * Dead code finding could be added with [yagni](https://github.com/venantius/yagni)

Reviewer: Dead code detection
